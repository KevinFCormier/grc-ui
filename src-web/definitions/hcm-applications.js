/*******************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2018. All Rights Reserved.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 *******************************************************************************/
'use strict'
import React from 'react'
import { Loading } from 'carbon-components-react'
import lodash from 'lodash'
import { getAge, getLabelsToString, getLabelsToList } from '../../lib/client/resource-helper'
import { MCM_OPEN_DIAGRAM_TAB_COOKIE } from '../../lib/shared/constants'
import msgs from '../../nls/platform.properties'
import { Link } from 'react-router-dom'
import config from '../../lib/shared/config'

export default {
  defaultSortField: 'metadata.name',
  uriKey: 'metadata.name',
  primaryKey: 'metadata.name',
  secondaryKey: 'metadata.namespace',
  tableKeys: [
    {
      msgKey: 'table.header.name',
      resourceKey: 'metadata.name',
      transformFunction: createApplicationLink,
    },
    {
      msgKey: 'table.header.namespace',
      resourceKey: 'metadata.namespace'
    },
    {
      msgKey: 'table.header.labels',
      resourceKey: 'metadata.labels',
      transformFunction: getLabelsToString
    },
    {
      msgKey: 'table.header.created',
      resourceKey: 'metadata.creationTimestamp',
      transformFunction: getAge,
    },
    // {
    //   msgKey: 'table.header.status',
    //   resourceKey: 'metadata.status',
    //   transformFunction: getStatus,
    // },
    {
      msgKey: 'table.header.dashboard',
      resourceKey: 'dashboard',
      transformFunction: createDashboardLink,
    },
  ],
  tableActions: [
    'table.actions.applications.remove',
  ],
  detailKeys: {
    title: 'application.details',
    headerRows: ['type', 'detail'],
    rows: [
      {
        cells: [
          {
            resourceKey: 'description.title.name',
            type: 'i18n'
          },
          {
            resourceKey: 'metadata.name'
          }
        ]
      },
      {
        cells: [
          {
            resourceKey: 'description.title.namespace',
            type: 'i18n'
          },
          {
            resourceKey: 'metadata.namespace'
          }
        ]
      },
      {
        cells: [
          {
            resourceKey: 'description.title.created',
            type: 'i18n'
          },
          {
            resourceKey: 'metadata.creationTimestamp',
            transformFunction: getAge
          }
        ]
      },
      // {
      //   cells: [
      //     {
      //       resourceKey: 'description.title.status',
      //       type: 'i18n'
      //     },
      //     {
      //       resourceKey: 'status',
      //       transformFunction: getStatus
      //     }
      //   ]
      // },
      {
        cells: [
          {
            resourceKey: 'description.title.labels',
            type: 'i18n'
          },
          {
            resourceKey: 'metadata.labels',
            transformFunction: getLabelsToList
          }
        ]
      },
      {
        cells: [
          {
            resourceKey: 'description.title.selector',
            type: 'i18n'
          },
          {
            resourceKey: 'selector',
            transformFunction: getLabelsToList,
          }
        ]
      },
      {
        cells: [
          {
            resourceKey: 'description.title.annotations',
            type: 'i18n'
          },
          {
            resourceKey: 'metadata.annotations',
            transformFunction: getLabelsToList,
          }
        ]
      },
      {
        cells: [
          {
            resourceKey: 'description.title.resource.version',
            type: 'i18n'
          },
          {
            resourceKey: 'metadata.resourceVersion'
          }
        ]
      },
      {
        cells: [
          {
            resourceKey: 'description.title.self.link',
            type: 'i18n'
          },
          {
            resourceKey: 'metadata.selfLink'
          }
        ]
      },
      {
        cells: [
          {
            resourceKey: 'description.title.uid',
            type: 'i18n'
          },
          {
            resourceKey: 'metadata.uid'
          }
        ]
      },
    ]
  },
  applicationWorkKeys: {
    title: 'application.works',
    defaultSortField: 'name',
    resourceKey: 'applicationWorks',
    normalizedKey: 'metadata.name',
    tableKeys: [
      {
        key: 'name',
        resourceKey: 'metadata.name',
        msgKey: 'table.header.name'
      },
      {
        key: 'namespace',
        resourceKey: 'metadata.namespace',
        msgKey: 'table.header.namespace'
      },
      {
        key: 'cluster',
        resourceKey: 'cluster',
        msgKey: 'table.header.cluster'
      },
      {
        key: 'release',
        resourceKey: 'release',
        msgKey: 'table.header.helm.release'
      },
      {
        key: 'status',
        resourceKey: 'status',
        msgKey: 'table.header.status'
      },
      {
        key: 'reason',
        resourceKey: 'reason',
        msgKey: 'table.header.reason'
      },
      {
        key: 'timestamp',
        resourceKey: 'metadata.creationTimestamp',
        msgKey: 'table.header.created',
        transformFunction: getAge
      },
    ],
  },
  placementPolicyKeys: {
    title: 'application.placement.policies',
    defaultSortField: 'name',
    resourceKey: 'placementPolicies',
    normalizedKey: 'metadata.name',
    tableKeys: [
      {
        key: 'name',
        resourceKey: 'metadata.name',
        msgKey: 'table.header.name'
      },
      {
        key: 'namespace',
        resourceKey: 'metadata.namespace',
        msgKey: 'table.header.namespace'
      },
      {
        key: 'replicas',
        resourceKey: 'clusterReplicas',
        msgKey: 'table.header.replicas'
      },
      {
        key: 'clusterSelector',
        resourceKey: 'clusterLabels',
        msgKey: 'table.header.cluster.selector',
        transformFunction: getLabelsToList,
      },
      {
        key: 'resourceSelector',
        resourceKey: 'resourceSelector',
        msgKey: 'table.header.resource.selector',
        transformFunction: getLabelsToList,
      },
      {
        key: 'decisions',
        resourceKey: 'status',
        msgKey: 'table.header.decisions',
        transformFunction: getDecisions,
      },
      {
        key: 'timestamp',
        resourceKey: 'metadata.creationTimestamp',
        msgKey: 'table.header.created',
        transformFunction: getAge
      },
    ],
    tableActions: [
      'table.actions.edit',
    ],
  },
  placementBindingKeys: {
    title: 'application.placement.bindings',
    defaultSortField: 'name',
    resourceKey: 'placementBindings',
    normalizedKey: 'metadata.name',
    tableKeys: [
      {
        key: 'name',
        resourceKey: 'metadata.name',
        msgKey: 'table.header.name'
      },
      {
        key: 'namespace',
        resourceKey: 'metadata.namespace',
        msgKey: 'table.header.namespace'
      },
      {
        key: 'placementpolicy',
        resourceKey: 'placementRef.name',
        msgKey: 'table.header.placementpolicy'
      },
      {
        key: 'subjects',
        resourceKey: 'subjects',
        msgKey: 'table.header.subjects',
        transformFunction: getSubjects
      },
      {
        key: 'timestamp',
        resourceKey: 'metadata.creationTimestamp',
        msgKey: 'table.header.created',
        transformFunction: getAge
      },
    ]
  },
  deployablesKeys: {
    title: 'application.deployables',
    defaultSortField: 'name',
    resourceKey: 'deployables',
    normalizedKey: 'metadata.name',
    tableKeys: [
      {
        key: 'name',
        resourceKey: 'metadata.name',
        msgKey: 'table.header.name'
      },
      {
        key: 'namespace',
        resourceKey: 'metadata.namespace',
        msgKey: 'table.header.namespace'
      },
      {
        key: 'chart',
        resourceKey: 'deployer',
        msgKey: 'table.header.deployerDetails',
        transformFunction: getDeployerDetails
      },
      {
        key: 'dependencies',
        resourceKey: 'dependencies',
        msgKey: 'table.header.dependencies',
        transformFunction: getDependencies
      },
      {
        key: 'timestamp',
        resourceKey: 'metadata.creationTimestamp',
        msgKey: 'table.header.created',
        transformFunction: getAge
      },
    ],
  },
  applicationRelationshipKeys: {
    title: 'application.relationships',
    defaultSortField: 'name',
    resourceKey: 'applicationRelationships',
    normalizedKey: 'metadata.name',
    tableKeys: [
      {
        key: 'name',
        resourceKey: 'metadata.name',
        msgKey: 'table.header.name'
      },
      {
        key: 'namespace',
        resourceKey: 'metadata.namespace',
        msgKey: 'table.header.namespace'
      },
      {
        key: 'source',
        resourceKey: 'source',
        msgKey: 'table.header.source',
        transformFunction: getRelationshipSourceDest
      },
      {
        key: 'destination',
        resourceKey: 'destination',
        msgKey: 'table.header.destination',
        transformFunction: getRelationshipSourceDest
      },
      {
        key: 'type',
        resourceKey: 'type',
        msgKey: 'table.header.type'
      },
      {
        key: 'timestamp',
        resourceKey: 'metadata.creationTimestamp',
        msgKey: 'table.header.created',
        transformFunction: getAge
      },
    ],
  },
}

export function createApplicationLink(item = {}, ...param){
  const {metadata: {name, namespace = 'default'}} = item
  if (param[2]) return item.metadata.name
  let link = `${config.contextPath}/applications/${encodeURIComponent(namespace)}/${encodeURIComponent(name)}`
  // if user at any point opens the diagram tab, they will continue to open application diagrams isntead of overview
  if (localStorage.getItem(MCM_OPEN_DIAGRAM_TAB_COOKIE)) {
    link = link+'/diagram'
  }
  return <Link to={link}>{name}</Link>
}

export function createDashboardLink({ dashboard = '' } , locale){
  if (dashboard !== null && dashboard !== '' )
    return <a target="_blank" rel="noopener noreferrer" href={dashboard}>{msgs.get('table.actions.launch.grafana', locale)}</a>

  return '-'
}

export function getStatus(item = {}){
  return item.hasPendingActions ?
    <Loading id={`loading-${item.name}`} small withOverlay={false} />
    :
    item.status
}

export function getDecisions(item = {}){
  const decisions = lodash.get(item, 'placementPolicies[0].status.decisions') || lodash.get(item, 'status.decisions')
  if (decisions) {
    return decisions.map(decision => decision.clusterName).join(', ')
  }
  return '-'
}

export function getDeployerDetails(item = {}, locale){
  if (item.deployer){
    // deployer was a chart
    if (item.deployer.chartURL) {
      return (
        <ul>
          <li key='name' style={{display:'block'}}>
            <b>{msgs.get('table.header.chartUrl', locale)}</b>{` = ${item.deployer.chartURL ? item.deployer.chartURL : '-'}`}
          </li>
          <li key='namespace' style={{display:'block'}}>
            <b>{msgs.get('table.header.chartNamespace', locale)}</b>{` = ${item.deployer.namespace ? item.deployer.namespace : '-'}`}
          </li>
        </ul>
      )
      // deployer was a chart
    } else if (item.deployer.chartName) {
      return (
        <ul>
          <li key='name' style={{display:'block'}}>
            <b>{msgs.get('table.header.chartName', locale)}</b>{` = ${item.deployer.chartName ? item.deployer.chartName : '-'}`}
          </li>
          <li key='repo' style={{display:'block'}}>
            <b>{msgs.get('table.header.chartRepo', locale)}</b>{` = ${item.deployer.repository ? item.deployer.repository : '-'}`}
          </li>
          <li key='version' style={{display:'block'}}>
            <b>{msgs.get('table.header.chartVersion', locale)}</b>{` = ${item.deployer.version ? item.deployer.version : '-'}`}
          </li>
          <li key='namespace' style={{display:'block'}}>
            <b>{msgs.get('table.header.chartNamespace', locale)}</b>{` = ${item.deployer.namespace ? item.deployer.namespace : '-'}`}
          </li>
        </ul>
      )
    } else if (item.deployer.kubeKind) {
      // deployer was a k8 object
      return (
        <ul>
          <li key='name' style={{display:'block'}}>
            <b>{msgs.get('table.header.kind', locale)}</b>{` = ${item.deployer.kubeKind ? item.deployer.kubeKind : '-'}`}
          </li>
          <li key='namespace' style={{display:'block'}}>
            <b>{msgs.get('table.header.name', locale)}</b>{` = ${item.deployer.kubeName ? item.deployer.kubeName : '-'}`}
          </li>
        </ul>
      )
    }
  }
}

export function getDependencies(item = {}){
  if (item.dependencies) {
    let str = ''
    item.dependencies.forEach(({name, kind}) => {
      str += `${name} [${kind}], `
    })
    return str.substring(0, str.length - 2)
  }
  return '-'
}

export function getRelationshipSourceDest(item, locale, arg) {
  return arg === 'source'
    ? <ul>
      <li style={{display:'block'}}><b>{'name'}</b>{` = ${item.source ? item.source.name : '-'}`}</li>
      <li style={{display:'block'}}><b>{'kind'}</b>{` = ${item.source ? item.source.name : '-'}`}</li>
    </ul>
    : <ul>
      <li style={{display:'block'}}><b>{'name'}</b>{` = ${item.destination ? item.destination.name : '-'}`}</li>
      <li style={{display:'block'}}><b>{'kind'}</b>{` = ${item.destination ? item.destination.name : '-'}`}</li>
    </ul>
}

export function getSubjects(item) {
  return item.subjects && item.subjects.map(subject => `${subject.name}(${subject.kind})`).join(', ')
}
