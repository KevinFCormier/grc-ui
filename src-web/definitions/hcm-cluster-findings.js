/*******************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2019. All Rights Reserved.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 *******************************************************************************/
'use strict'
import _ from 'lodash'

export default {
  defaultSortField: 'cluster',
  primaryKey: 'cluster',
  secondaryKey: 'metadata.namespace',
  tableActions: [
    'table.actions.finding.sidepanel',
  ],
  tableKeys: [
    {
      msgKey: 'table.header.cluster.name',
      resourceKey: 'cluster',
    },
    {
      msgKey: 'table.header.namespace',
      resourceKey: 'namespace',
    },
    {
      msgKey: 'table.header.severity',
      resourceKey: 'severity',
    },
    {
      msgKey: 'table.header.highSeverity',
      resourceKey: 'highSeverity',
      transformFunction: getHighSeverityObjTostring,
    },
  ],
}

export function getHighSeverityObjTostring(item) {
  const highSeverityMap = _.get(item, 'highSeverity')
  let highSeverityString = ''
  if (highSeverityMap) {
    Object.keys(highSeverityMap).forEach((key) => {
      highSeverityString = `${highSeverityString} ${key} (${highSeverityMap[key]}),`
    })
  }
  highSeverityString = highSeverityString.trim()
  return highSeverityString.substring(0, highSeverityString.length-1)
}
