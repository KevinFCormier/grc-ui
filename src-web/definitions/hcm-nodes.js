/*******************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2018. All Rights Reserved.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 *******************************************************************************/
'use strict'
import { getAge } from '../../lib/client/resource-helper'

export default {
  defaultSortField: 'cluster',
  primaryKey: 'metadata.name',
  secondaryKey: 'cluster.metadata.name',
  tableKeys: [
    {
      msgKey: 'table.header.name',
      resourceKey: 'metadata.name',
    },
    {
      msgKey: 'table.header.cluster',
      resourceKey: 'cluster.metadata.name',
    },
    {
      msgKey: 'table.header.role',
      resourceKey: 'metadata.labels',
      transformFunction: getRole
    },
    {
      msgKey: 'table.header.arch',
      resourceKey: 'architecture',
    },
    {
      msgKey: 'table.header.osimage',
      resourceKey: 'osImage',
    },
    {
      msgKey: 'table.header.cpus',
      resourceKey: 'capacity.cpu',
    },
    {
      msgKey: 'table.header.created',
      resourceKey: 'metadata.creationTimestamp',
      transformFunction: getAge,
    },
  ],
}

export function getRole(item) {
  return (item.roles && item.roles.length > 0) ? item.roles.join(', ') : 'worker'
}
