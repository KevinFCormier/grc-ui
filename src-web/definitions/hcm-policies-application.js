/*******************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2019. All Rights Reserved.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 *******************************************************************************/
'use strict'

import React from 'react'
import TruncateText from '../components/common/TruncateText'

export default {
  defaultSortField: 'name',
  primaryKey: 'name',
  secondaryKey: 'namespace',
  tableActions: [
    'table.actions.policy.applications.sidepanel',
  ],
  tableKeys: [
    {
      msgKey: 'table.header.application',
      resourceKey: 'name',
    },
    {
      msgKey: 'table.header.namespace',
      resourceKey: 'nameSpace',
    },
    {
      msgKey: 'table.header.violation',
      resourceKey: 'violations',
    },
    {
      msgKey: 'table.header.violated',
      resourceKey: 'violatedPolicies',
      transformFunction: getLabels,
    }
  ],
  applicationViolatedSidePanel: {
    headerRows: ['', 'table.header.policy.name', 'table.header.rule.violation', 'table.header.control', 'table.header.cluster',],
    subHeaders: ['table.header.name', 'table.header.message', 'table.header.reason', '',],
    rows: [
      {
        cells: [
          {
            resourceKey: 'metadata.name',
          },
          {
            resourceKey: 'violatedNum',
          },
          {
            resourceKey: 'metadata.annotations["policy.mcm.ibm.com/controls"]',
          },
          {
            resourceKey: 'cluster',
          },
        ]
      }
    ]
  },
}

export function getLabels(item) {
  const violatedPoliciesName = []
  item.violatedPolicies.forEach(policy => {
    violatedPoliciesName.push(policy.name)
  })
  return getTruncatedText(violatedPoliciesName.join(', '))
}

export function getTruncatedText(item){
  return <TruncateText text={item} />
}
