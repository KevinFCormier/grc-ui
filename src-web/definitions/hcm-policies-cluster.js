/*******************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2018. All Rights Reserved.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 *******************************************************************************/
'use strict'

import React from 'react'
import TruncateText from '../components/common/TruncateText'

export default {
  defaultSortField: 'cluster',
  primaryKey: 'cluster',
  secondaryKey: 'metadata.namespace',
  tableActions: [
    'table.actions.sidepanel',
  ],
  tableKeys: [
    {
      msgKey: 'table.header.cluster.name',
      resourceKey: 'cluster',
    },
    {
      msgKey: 'table.header.violation',
      resourceKey: 'violation',
    },
    {
      msgKey: 'table.header.violated',
      resourceKey: 'nonCompliant',
      transformFunction: getLabels,
    }
  ],
  violatedTable: {
    // title: 'cluster.violated.policy',
    headerRows: ['', 'table.header.policy.name', 'table.header.rule.violation', 'table.header.control'],
    subHeaders: ['table.header.name', 'table.header.message', 'table.header.reason'],
    rows: [
      {
        cells: [
          {
            resourceKey: 'metadata.name',
          },
          {
            resourceKey: 'violatedNum',
          },
          {
            resourceKey: 'metadata.annotations["policy.mcm.ibm.com/controls"]',
          }
        ]
      }
    ]
  },
}

export function getLabels(item) {
  return getTruncatedText(item.nonCompliant.join(', '))
}

export function getTruncatedText(item){
  return <TruncateText text={item} />
}
