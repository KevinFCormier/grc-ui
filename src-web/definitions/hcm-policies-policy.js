/*******************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2018. All Rights Reserved.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 *******************************************************************************/
'use strict'

import {getCategories, getControls, getStandards} from './hcm-compliances'

export default {
  defaultSortField: 'metadata.name',
  primaryKey: 'metadata.name',
  secondaryKey: 'metadata.namespace',
  tableKeys: [
    {
      msgKey: 'table.header.name',
      resourceKey: 'metadata.name',
    },
    {
      msgKey: 'table.header.namespace',
      resourceKey: 'metadata.namespace',
    },
    {
      msgKey: 'table.header.remediation',
      resourceKey: 'remediation',
    },
    {
      msgKey: 'table.header.cluster.compliant',
      resourceKey: 'clusterCompliant',
    },
    {
      msgKey: 'table.header.controls',
      resourceKey: 'metadata.annotations["policy.mcm.ibm.com/controls"]',
      transformFunction: getControls,
    },
    {
      msgKey: 'table.header.standards',
      resourceKey: 'metadata.annotations["policy.mcm.ibm.com/standards"]',
      transformFunction: getStandards,
    },
    {
      msgKey: 'table.header.categories',
      resourceKey: 'metadata.annotations["policy.mcm.ibm.com/categories"]',
      transformFunction: getCategories
    },
  ],
}

