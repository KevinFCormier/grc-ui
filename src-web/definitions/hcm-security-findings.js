/*******************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2019. All Rights Reserved.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 *******************************************************************************/
'use strict'

import _ from 'lodash'
import { getAge } from '../../lib/client/resource-helper'

export default {
  defaultSortField: 'shortDescription',
  primaryKey: 'name',
  secondaryKey: 'finding.severity',
  tableKeys: [
    {
      msgKey: 'table.header.description',
      resourceKey: 'shortDescription',
    },
    {
      msgKey: 'table.header.resources',
      resourceKey: 'context',
      transformFunction: showTypeAndName
    },
    {
      msgKey: 'table.header.severity',
      resourceKey: 'finding.severity',
    },
    {
      msgKey: 'table.header.cluster',
      resourceKey: 'context.clusterName',
    },
    {
      msgKey: 'table.header.standards',
      resourceKey: 'securityClassification.securityStandards',
      transformFunction: getFindingStandards
    },
    {
      msgKey: 'table.header.controls',
      resourceKey: 'securityClassification.securityControl',
      transformFunction: getFindingControl
    },
    {
      msgKey: 'table.header.categories',
      resourceKey: 'securityClassification.securityCategories',
      transformFunction: getFindingCategories
    },
    {
      msgKey: 'table.header.update.time',
      resourceKey: 'updateTime',
      transformFunction: getAge
    },
  ],
}

function compressArray(items){
  if (items) {
    return items.map(item => _.startCase(item.trim())).join(', ')
  }
  return '-'
}

function showTypeAndName(item){
  const resourceType = _.get(item, 'context.resourceType', '-')
  const resourceName = _.get(item, 'context.resourceName', '-')
  return `${resourceType}: ${resourceName}`
}

export function getFindingStandards(item) {
  const securityStandards = compressArray(_.get(item, 'securityClassification.securityStandards') || [])
  return securityStandards ? securityStandards : '-'
}

export function getFindingControl(item) {
  const securityControl = _.startCase((_.get(item, 'securityClassification.securityControl') || '').trim())
  return securityControl ? securityControl : '-'
}

export function getFindingCategories(item) {
  const securityCategories = compressArray(_.get(item, 'securityClassification.securityCategories') || [])
  return securityCategories ? securityCategories : '-'
}
