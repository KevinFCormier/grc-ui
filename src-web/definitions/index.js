/*******************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2019. All Rights Reserved.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 *******************************************************************************/
'use strict'

import lodash from 'lodash'

import { RESOURCE_TYPES } from '../../lib/shared/constants'

import hcmpoliciescluster from './hcm-policies-cluster'
import hcmpoliciespolicy from './hcm-policies-policy'
import hcmcompliances from './hcm-compliances'
import hcmpolicies from './hcm-policies'


const resourceData = {
  [RESOURCE_TYPES.HCM_POLICIES_PER_POLICY.name]: hcmpoliciespolicy,
  [RESOURCE_TYPES.HCM_POLICIES_PER_CLUSTER.name]: hcmpoliciescluster,
  [RESOURCE_TYPES.HCM_COMPLIANCES.name]: hcmcompliances,
  [RESOURCE_TYPES.HCM_POLICIES.name]: hcmpolicies,
}

// diagram tabs

function getResourceData(resourceType) {
  // main Topology tab
  const def = resourceData[resourceType.name]
  if (!def) {
    //eslint-disable-next-line no-console
    console.error(`No resource data found for '${resourceType}'`)
  }
  return def
}

export default getResourceData

export function getPrimaryKey(resourceType) {
  const def = getResourceData(resourceType)
  let pk = def.primaryKey

  if (!pk)
    pk = 'metadata.uid'

  return pk
}

export function getSecondaryKey(resourceType) {
  const def = getResourceData(resourceType)
  let sk = def.secondaryKey

  if (!sk)
    sk = 'cluster'

  return sk
}

export function getURIKey(resourceType) {
  const def = getResourceData(resourceType)
  let uriKey = def.uriKey

  if (!uriKey)
    uriKey = 'metadata.name'

  return uriKey
}

export function getDefaultSearchField(resourceType) {
  var def = getResourceData(resourceType)
  let sf = def && def.defaultSearchField
  if (!def || !def.tableKeys || (def && def.tableKeys.length < 1))
    //eslint-disable-next-line no-console
    console.error(`No table keys found in ${resourceType} resource definition`)
  if (!sf)
    sf = def && def.tableKeys && def.tableKeys[0] && def.tableKeys[0].resourceKey
  return sf
}

export function getDefaultSortField(resourceType) {
  var def = getResourceData(resourceType)
  let sf = def && def.defaultSortField
  if (!def || !def.tableKeys || (def && def.tableKeys.length < 1))
    //eslint-disable-next-line no-console
    console.error(`No table keys found in ${resourceType} resource definition`)
  if (!sf) {
    sf = def && def.tableKeys && def.tableKeys[0] && def.tableKeys[0].resourceKey
  }
  if (!sf)
    //eslint-disable-next-line no-console
    console.error(`No sortable fields defined for '${resourceType}' resource definition`)
  return sf
}

export function getTableKeys(resourceType) {
  var def = getResourceData(resourceType)
  return def.tableKeys
}

export function getLink(link, resource) {
  if(typeof(link) === 'boolean') {
    return `/${resource.metadata.namespace}/${resource.metadata.name}`
  } else if (typeof(link) === 'function') {
    return link(resource)
  } else {
    const parts = link.split('/')
    let path = ''
    parts.forEach(part => {
      const value = lodash.get(resource, part)
      path += `/${encodeURIComponent(value)}`
    })
    return path
  }
}
