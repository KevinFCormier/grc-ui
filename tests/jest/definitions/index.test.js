/*******************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2018, 2019. All Rights Reserved.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 *******************************************************************************/
'use strict'
import {getDefaultSearchField, getDefaultSortField, getLink, getPrimaryKey, getSecondaryKey } from '../../../src-web/definitions/index'

// regular properties/selector/reducer testing
// compare the received and expect values
describe('definitions/index', () => {
  describe('#getDefaultSearchField', () => {
    it('should return the default search field of node', () => {
      const item = {
        list: 'HCMPolicyList',
        name: 'HCMPolicy'
      }
      expect(getDefaultSearchField(item)).toBe('metadata.name')
    })
  })

  describe('#getDefaultSortField', () => {
    it('should return the default sort field of node', () => {
      const item = {
        list: 'HCMPolicyList',
        name: 'HCMPolicy'
      }
      expect(getDefaultSortField(item)).toBe('metadata.name')
    })
  })

  describe('#getLink', () => {
    it('should return the link of node as /namespace/name', () => {
      const link = true
      const resource = {
        metadata: {
          namespace: 'namespace',
          name: 'name'
        }
      }
      expect(getLink(link, resource)).toBe('/namespace/name')
    })
    it('should return the link of node as /domain/host', () => {
      const link = 'domain/host'
      const resource = {
        domain: 'domain',
        host: 'host',
        metadata: {
          namespace: 'namespace',
          name: 'name'
        }
      }
      expect(getLink(link, resource)).toBe('/domain/host')
    })
  })

  describe('#getPrimaryKey', () => {
    it('should return the primary key of node', () => {
      const item = {
        list: 'HCMPolicyList',
        name: 'HCMPolicy'
      }
      expect(getPrimaryKey(item)).toBe('metadata.name')
    })
  })

  describe('#getSecondaryKey', () => {
    it('should return the secondary key of node', () => {
      const item = {
        list: 'HCMPolicyList',
        name: 'HCMPolicy'
      }
      expect(getSecondaryKey(item)).toBe('metadata.namespace')
    })
  })

  describe('#getURIKey', () => {
    //getURIKey not been used
  })
})
