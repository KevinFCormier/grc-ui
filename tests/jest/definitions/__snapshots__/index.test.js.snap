// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
Array [
  Object {
    "msgKey": "table.header.description",
    "resourceKey": "shortDescription",
  },
  Object {
    "msgKey": "table.header.resources",
    "resourceKey": "context",
    "transformFunction": [Function],
  },
  Object {
    "msgKey": "table.header.severity",
    "resourceKey": "finding.severity",
    "transformFunction": [Function],
  },
  Object {
    "msgKey": "table.header.cluster",
    "resourceKey": "context.clusterName",
  },
  Object {
    "msgKey": "table.header.standards",
    "resourceKey": "securityClassification.securityStandards",
    "transformFunction": [Function],
  },
  Object {
    "msgKey": "table.header.controls",
    "resourceKey": "securityClassification.securityControl",
    "transformFunction": [Function],
  },
  Object {
    "msgKey": "table.header.categories",
    "resourceKey": "securityClassification.securityCategories",
    "transformFunction": [Function],
  },
  Object {
    "msgKey": "table.header.update.time",
    "resourceKey": "updateTime",
    "transformFunction": [Function],
  },
]
`;
